testdata/p4tc_samples_outputs/add_entry_1_example_control_blocks.c

```diff
@@ -0,0 +1,315 @@
+};
+
+static __always_inline int process(struct __sk_buff *skb, struct headers_t *hdr, struct pna_global_metadata *compiler_meta__)
+{
+    struct hdr_md *hdrMd;
+
+    unsigned ebpf_packetOffsetInBits_save = 0;
+    ParserError_t ebpf_errorCode = NoError;
+    void* pkt = ((void*)(long)skb->data);
+    u8* hdr_start = pkt;
+    void* ebpf_packetEnd = ((void*)(long)skb->data_end);
+    u32 ebpf_zero = 0;
+    u32 ebpf_one = 1;
+    unsigned char ebpf_byte;
+    u32 pkt_len = skb->len;
+
+    struct main_metadata_t *user_meta;
+    hdrMd = BPF_MAP_LOOKUP_ELEM(hdr_md_cpumap, &ebpf_zero);
+    if (!hdrMd)
+        return TC_ACT_SHOT;
+    unsigned ebpf_packetOffsetInBits = hdrMd->ebpf_packetOffsetInBits;
+    hdr_start = pkt + BYTES(ebpf_packetOffsetInBits);
+    hdr = &(hdrMd->cpumap_hdr);
+    user_meta = &(hdrMd->cpumap_usermeta);
+{
+        u8 hit;
+        {
+if (/* hdr->ipv4.isValid() */
+            hdr->ipv4.ebpf_valid) {
+/* ipv4_tbl.apply() */
+                {
+                    /* construct key */
+                    struct p4tc_table_entry_act_bpf_params__local params = {
+                        .pipeid = p4tc_filter_fields.pipeid,
+                        .tblid = 1
+                    };
+                    struct MainControlImpl_ipv4_tbl_1_key key = {};
+                    key.keysz = 64;
+                    key.field0 = hdr->ipv4.dstAddr;
+                    key.field1 = skb->ifindex;
+                    struct p4tc_table_entry_act_bpf *act_bpf;
+                    /* value */
+                    struct MainControlImpl_ipv4_tbl_1_value *value = NULL;
+                    /* perform lookup */
+                    act_bpf = bpf_p4tc_tbl_read(skb, &params, &key, sizeof(key));
+                    value = (struct MainControlImpl_ipv4_tbl_1_value *)act_bpf;
+                    if (value == NULL) {
+                        /* miss; find default action */
+                        hit = 0;
+                    } else {
+                        hit = 1;
+                    }
+                    if (value != NULL) {
+                        /* run action */
+                        switch (value->action) {
+                            case MAINCONTROLIMPL_IPV4_TBL_1_ACT_MAINCONTROLIMPL_NEXT_HOP: 
+                                {
+/* add_entry(""send_nh"", {hdr->ethernet.srcAddr, hdr->ethernet.dstAddr}, 2) */
+                                    struct p4tc_table_entry_act_bpf update_act_bpf = {};
+                                    struct act_param {
+                                        u64 param0;
+                                        u64 param1;
+                                    };
+                                    struct act_param* params = (struct act_param *) update_act_bpf.params;
+                                    params->param0 = hdr->ethernet.srcAddr;
+                                    params->param1 = hdr->ethernet.dstAddr;
+                                    update_act_bpf.act_id = 1;
```

Comment:
It would be simpler (and the verifier wouldn't complain much) if you use `struct MainControlImpl_ipv4_tbl_1_value` directly instead creating `struct act_param`.
So something like this:

```suggestion
                                    struct p4tc_table_entry_act_bpf update_act_bpf = {};
                                    struct MainControlImpl_ipv4_tbl_1_value *update_act_bpf_val =
                                            (struct MainControlImpl_ipv4_tbl_1_value *)&update_act_bpf;
                                    update_act_bpf_val->u.MainControlImpl_send_nh.smac = hdr->ethernet.srcAddr;
                                    update_act_bpf_val->u.MainControlImpl_send_nh.dmac = hdr->ethernet.dstAddr;
                                    update_act_bpf_val.action = MAINCONTROLIMPL_IPV4_TBL_1_ACT_MAINCONTROLIMPL_SEND_NH;
```

Reply:
addressed